### PR TITLE
rounded rect

### LINKED ISSUE #2054: Rounded corners on rect, but only for some corners
As originally suggested in #1201, this is a common request, and the workaround isn’t always feasible (for example with diverging bars).

image

### PR DESCRIPTION
**Pull request by mbostock**
TODO

- [x] rect
- [x] bar & cell
- [x] frame
- [x] **r** option
- [x] handle inverted orientations
- [x] handle corner radii that are too big
- [x] treat negative or invalid radii as zero
- [x] handle insets
- [x] handle missing channels
- [x] handle collapsed channels
- [x] handle projections
- [x] allow larger corners if asymmetric
- [x] consolidate code between rect, bar, and cell
- [x] optimize conditionals?
- [x] documentation
- [x] tests

Fixes #2054.

### CODE REVIEW

File: test/output/roundedRectAsymmetricX.svg
Comment (Fil) [inline test/output/roundedRectAsymmetricX.svg:1]: NICE!

**Comment by mbostock** [comment]
The implementation is now feature-complete and I’d welcome testing or feedback. I might try to optimize the conditionals, and add more documentation, before marking this as ready.

**Comment by mbostock** [comment]
I believe this is documented and ready for review. PTAL @Fil!

File: docs/marks/rect.md [hunk 1]
```diff
@@ -186,6 +186,62 @@ Plot.plot({
 A similar chart could be made with the [cell mark](./cell.md) using ordinal *x* and *y* scales instead, or with the [dot mark](./dot.md) as a scatterplot.
 :::
 
+To round corners, use the **r** option.<a id="r" href="#r" aria-label="Permalink to &quot;r&quot;"></a> If the combined corner radii excede the width or height of the rect, the radii are proportionally reduced to produce a pill shape with circular caps. Try increasing the radii below.
```
Comment (Fil) [inline docs/marks/rect.md:189]:
```suggestion
To round corners, use the **r** option.<a id="r" href="#r" aria-label="Permalink to &quot;r&quot;"></a> If the combined corner radii exceed the width or height of the rect, the radii are proportionally reduced to produce a pill shape with circular caps. Try increasing the radii below.
```

**Review by Fil — APPROVED** [review]
💊

**Pull request merged**